controller/volume_controller.go

```diff
@@ -3332,3 +3469,85 @@ func (vc *VolumeController) ReconcileBackupVolumeState(volume *longhorn.Volume)
 	volume.Status.LastBackupAt = bv.Status.LastBackupAt
 	return nil
 }
+
+// TODO: this block of code is duplicated of CreateSnapshot in MANAGER package.
+// Once we have Snapshot CR, we should refactor this
+
+func (vc *VolumeController) createSnapshot(snapshotName string, labels map[string]string, volume *longhorn.Volume) (*types.Snapshot, error) {
+	if volume.Name == "" {
+		return nil, fmt.Errorf("volume name required")
+	}
+
+	for k, v := range labels {
+		if strings.Contains(k, "=") || strings.Contains(v, "=") {
+			return nil, fmt.Errorf("labels cannot contain '='")
+		}
+	}
+
+	if err := vc.checkVolumeNotInMigration(volume); err != nil {
+		return nil, err
+	}
+
+	engine, err := vc.getEngineClient(volume.Name)
+	if err != nil {
+		return nil, err
+	}
+	snapshotName, err = engine.SnapshotCreate(snapshotName, labels)
+	if err != nil {
+		return nil, err
+	}
+	snap, err := engine.SnapshotGet(snapshotName)
+	if err != nil {
+		return nil, err
+	}
+	if snap == nil {
+		return nil, fmt.Errorf("cannot found just created snapshot '%s', for volume '%s'", snapshotName, volume.Name)
+	}
+	logrus.Debugf("Created snapshot %v with labels %+v for volume %v", snapshotName, labels, volume.Name)
+	return snap, nil
+}
+
+func (vc *VolumeController) getEngineClient(volumeName string) (client engineapi.EngineClient, err error) {
+	var e *longhorn.Engine
+
+	defer func() {
+		err = errors.Wrapf(err, "cannot get client for volume %v", volumeName)
+	}()
+	es, err := vc.ds.ListVolumeEngines(volumeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(es) == 0 {
+		return nil, fmt.Errorf("cannot find engine")
+	}
+	if len(es) != 1 {
+		return nil, fmt.Errorf("more than one engine exists")
+	}
+	for _, e = range es {
+		break
+	}
+	if e.Status.CurrentState != types.InstanceStateRunning {
+		return nil, fmt.Errorf("engine is not running")
+	}
```

Comment:
You can use `vc.ds.PickVolumeCurrentEngine` instead

Reply:
I see that it doesn't save the mount of code. That func might return `nil` so we need to check for `nil`

Reply:
It makes the code more readable a little bit. But this is not required. I will mark this as resolved.